File: frontend/src/api/substitutionLog.ts
// 今日调课与代课看板与调课与代课日志(M4-4)。

import { apiGet } from '@/api/client'

export interface LogEntry {
  affected_period_id: number
  date: string
  weekday: number
  period_no: number
  period_name: string
  start_time: string | null
  end_time: string | null
  class_names: string
  subject_name: string
  room_name: string
  absent_teacher_id: number
  absent_teacher_name: string
  leave_type: string
  leave_type_label: string
  status: string
  status_label: string
  disposed: boolean
  sub_type: string | null
  sub_type_label: string | null
  handler_teacher_id: number | null
  handler_name: string | null
  counts_toward_hours: boolean | null
  swap_date: string | null
  swap_period_name: string
  swap_class_names: string
  swap_subject_name: string
  note: string
}

export interface DailyBoard {
  date: string
  weekday: number
  school_name: string
  semester_label: string
  entries: LogEntry[]
}

export interface LogFilters {
  teacherId?: number | null
  dateFrom?: string | null
  dateTo?: string | null
  leaveType?: string | null
}

export const getDailyBoard = (semesterId: number, on?: string | null): Promise<DailyBoard> =>
  apiGet(`/daily-board?semester_id=${semesterId}` + (on ? `&on=${on}` : ''))

export const getSubstitutionLog = (
  semesterId: number, f: LogFilters = {},
): Promise<LogEntry[]> => {
  const p = new URLSearchParams({ semester_id: String(semesterId) })
  if (f.teacherId) p.set('teacher_id', String(f.teacherId))
  if (f.dateFrom) p.set('date_from', f.dateFrom)
  if (f.dateTo) p.set('date_to', f.dateTo)
  if (f.leaveType) p.set('leave_type', f.leaveType)
  return apiGet(`/substitution-log?${p.toString()}`)
}

File: backend/app/schemas/substitution_log.py
"""今日看板与调课与代课日志 schema(M4-4)。"""

from datetime import date, time

from pydantic import BaseModel

# 字段名 date/start_time 会遮蔽同名类型,故以别名标注类型
_Date = date
_Time = time


class LogEntryOut(BaseModel):
    affected_period_id: int
    date: _Date
    weekday: int
    period_no: int
    period_name: str
    start_time: _Time | None = None
    end_time: _Time | None = None
    class_names: str
    subject_name: str
    room_name: str
    absent_teacher_id: int
    absent_teacher_name: str
    leave_type: str
    leave_type_label: str
    status: str
    status_label: str
    disposed: bool
    sub_type: str | None = None
    sub_type_label: str | None = None
    handler_teacher_id: int | None = None
    handler_name: str | None = None
    counts_toward_hours: bool | None = None
    swap_date: _Date | None = None
    swap_period_name: str = ""
    swap_class_names: str = ""
    swap_subject_name: str = ""
    note: str = ""


class DailyBoardOut(BaseModel):
    """今日看板:表头(校名/日期/学期)供打印通知单直接使用。"""

    date: _Date
    weekday: int
    school_name: str
    semester_label: str
    entries: list[LogEntryOut] = []
